# Lab Exercises

## Introduction to Python, Part 1

### Questions/Office Hours

If you have questions about this week's lab **ANY** TA for this class can help you with any problems in office hours. That being said, if you need help over email it is best to email **YOUR** TA.

You can also always ask the instructor for help with skills labs (even though they are not covered in lecture). You can email the instructor of come to his office hours.

----

## Exercises

These are primarily drawn from your reading for the week, chapters 2 and 3 of *Python for Everybody*. I've remixed and extended them a bit, so blame any errata on me, not Dr. Chuck. 

* [Click here to read Chapter 2 - Variables](https://www.py4e.com/html3/02-variables)
* [Click here to read Chapter 3 - Conditionals](https://www.py4e.com/html3/03-conditional)

### A brief note on these lab assignments

You're definitely going to get stuck at some point doing this assignment. When you do, there's a methodology that I think is helpful that you should try to develop. It comes from FreeCodeCamp (incidentally, a great resource for teaching yourself some coding), and it's called [Read-Search-Ask](https://medium.freecodecamp.org/read-search-dont-be-afraid-to-ask-743a23c411b4).

These three steps will get you through these assignments, and the habit will help you throughout your life coding. They're basically as follows:

1. *Read.* Reread the error messages you get, reread the exercise description, read the relevant parts from the textbook, read over any code that might be given to you, read over any supplementary materials you might've gotten that relate to what you're doing. Try any solutions or potential changes that come to mind after reading or were suggested while you're reading. You can do that as many times as you want; it doesn't hurt anybody to try putting in another parenthesis and hit Shift-Enter again.
2. *Search.* By now, you probably have a pretty good idea of what your problem is. Google it! Look on StackExchange, coding forums, and other websites. Don't copy and paste code (you know what plagiarism is) but try to see what others did that's different from what you have.
3. *Ask.* If you're still in lab and you're stuck, raise your hand! If not, drop me or the instructor an email. Come to either of our office hours. We'll help you out.


Read-Search-Ask is going to be helpful; keep it in the back of your mind as you're doing these assignments.

----

#### **Exercise 1**

###### sections to reference: 2.1, 2.4

Using the code cell directly below and the `type()` command, tell me the *type* of the following items:

* `184`
* `'The quick brown fox jumps over the lazy dog.'`
* `63.0`
* `'October 1917'`
* `'1776'`
* `'83.4'`
* `83.4`
* `'six'`
* `True`
* `False`
* `"True"`

You can leave your code in the cell if you want! For your answers, insert a Markdown cell below your code and make a list of the items and their type.

In [15]:
type(True)

bool

1. int
2. str
3. float
4. str
5. str
6. str
7. float
8. str
9. bool
10. bool
11. str

#### **Exercise 2**

###### sections to reference: 2.2, 2.3

Are the following items valid variable names in Python? If they are, just type "yes" next to the item on the list. If not, in a few words, explain why not.

* my_fun_variable
* item84
* 1600pennsylvania
* whereru@
* the_class
* class
* True
* val_true

In [24]:
val_true= "hello"

1. Yes
2. Yes
3. No, invalid decimal literal
4. No, unsupported operand type
5. Yes
6. No, reserved word
7. No, reserved word
8. Yes

#### **Exercise 3**

###### sections to reference: 2.4, 2.5, 2.6

Below are a list of, well, math problems. (Sorry.) First, go through the list and **write what you think it will evaluate to** in Python. Create a markdown cell and **make a table** to keep track of the math problems.

Then, using the code cell below this or the Python console in JupyterLabs (go to Launcher and click the "Python 3" box under "Console") below the list and **check your work**: type in the problems and see what output you get!

Finally, update your markdown table with the correct values in their own column. Then, below your table, **write** if any of the problems had different answers than you wrote in the beginning. Why do you think Python interpreted the math problems differently than you, a human? If all of your predictions were accurate, come up with an example or two of a problem where the result you think you *should* get isn't the result that Python gives you.

Here are the test math problems! Predict what you think they'll evaluate to, test them out, and write about what happened!

* `25+75`
* `52-8`
* `20*5`
* `55+9`
* `55.0+9`
* `55.0+9.0`
* `18.5-2`
* `61.2-3.2`
* `5+15+20`
* `8*20.0`
* `50/2`
* `50/2.0`
* `50//2.0`
* `50//2`
* `2**2`
* `4**0.5`
* `2**2**2`
* `5==5`
* `6==5`
* `8!=9`
* `42.0 > 30`
* `800 <= 800`

In [45]:
num1=50
num2=2
num3=20
sum=num1+num2
sum2=num1+num2+num3
diff=num1-num2
mul=num1*num2
div2= num1//num2
div= num1/num2
print(div2)

25


| Problem | Guess | Result |
|:--------:|:--------:|:--------:|
|  25+75   |  100   |  100   |
|  52-8   |  44  |  44   |
|  20*5  |  100   |  100  |
|  55+9  |  64   |  64   |
|  55.0+9 |  64.0  |  64.0   |
|  55.0+9.0 |  64.0   |  64.0  |
|  18.5-2   |  16.5   |  16.5   |
|  61.2-3.2  |  58.0  |  58.0   |
|  5+15+20 |  40   |  40  |
|  8*20.0  |  160.0   |  160.0    |
|  50/2 | 25  |  25.0  | (I thought it would not have a .0)
|  50//2.0 |  25.0   |  25.0  |
|  50//2  |  25   |  25   | (I know how that words now!)
|  2**2   |  4  |  4   |
|  4**0.5  |  2.0   |  2.0  |
|  2**2**2 |  8   |  16   | (2*2^2, cool)
|  5==5 |  true  |  true   |
|  6==5  false   |  false  |
|  8!=9   |  true   |  true   |
|  42.0 > 30  |  true  |  true   |
|  800 <= 800 |  true   |  true  |

#### **Exercise 4**

###### sections to reference: 2.5, 2.10

In the code cell below, write a program that asks the users for how many hours they work, their hourly wage, and what percent of their income they pay in taxes. Then, **print out** both their gross and net pay. Gross pay is what they made *before* taxes, and net pay is what they took home *after* taxes.

Hint: When you ask someone for the percent of income they pay in taxes, they're going to enter an integer like 30. In order to figure out what portion of their income they pay in taxes, you're going to need to convert that into a decimal between 0 and 1, so you can multiply it by their gross pay. How would you turn 30 into 0.3? Do that for the tax percentage they enter before you calculate their net pay.

In [49]:
hours = input("How many hours do you work?")
wage = input("What is your wage?")
taxes = input("What percent of your income they pay in taxes?")
hours= (float(hours))
wage= (float(wage))
taxes=(float(taxes))
gross=hours*wage
print(gross)
print(gross*taxes)


How many hours do you work? 15
What is your wage? 15
What percent of your income they pay in taxes? 10


225.0
2250.0


#### **Exercise 5**

###### sections to reference: 2.2, 2.3, 2.5, 2.6

This one is pretty similar to number 2, but with some complications. Let's say we define the following variables:

In [51]:
length = 20
width = 42.0
print(width//9)
print(width/9.0)
print(length/3)
print(1 + 4 * length)


4.0
4.666666666666667
6.666666666666667
81


If you happened to have cleared all inputs from this notebook, run the above code cell to make sure the variables are assigned.

Now, what will be the results of the following statements? Like before, **write** your answers next to the statements on the list, and then **check** them using a code cell or the Python console. Add a Markdown cell with your answers and if you get any wrong, tell me why!

* `width//9`
* `width/9.0`
* `length/3`
* `1 + 4 * length`

 Problem | Guess | Result |
|:--------:|:--------:|:--------:|
|  width//9   |  4.0   |  4.0   |
|  width/9.0  |  4.666666666666667  |  4.666666666666667   |
|  length/3  |  6.666666666666667  |  6.666666666666667  |
|  1 + 4 * length  |  81   |  81   | pretty acurate this time

#### **Exercise 6**

###### sections to reference: 2.5, 2.10, 3.1, 3.3, 3.4, 3.5

Write a program which prompts the user for a **temperature** and then asks if it was Celsius or Fahrenheit. If it was Celsius, **convert it** to Fahrenheit; if Fahrenheit, convert to Celsius. Then, **print** the converted temperature. 

I'm not going to give you the conversion formula here, so you'll have to search for it and then implement it in code. Also, you don't have to ask the user to type in "Celsius" or "Fahrenheit"; your prompt can be as simple as "Enter 1 if this temperature was Fahrenheit, or 0 if it was Celsius." That way, you're able to do something like `if isFahrenheit == 0` rather than exact string matching.

In [70]:
temp=input("What is the temperature")
temp2= input("Was that in Celsius (enter 1) or Fahrenheit (enter 0)")
if (int(temp2))==1:
    print(((float(temp)) * 9/5) +32)
else:
    print(float(temp))



What is the temperature 56
Was that in Celsius (enter 1) or Fahrenheit (enter 0) 1


132.8


#### **Exercise 7**

###### sections to reference: 2.8, 2.10, 3.1, 3.3, 3.4

This one's simple. Write a program that prompts the user for an integer. If the integer is even, print the integer with a message indicating that it's even. If it's odd, print it with a message saying that it's odd.

There are a few ways to do this, so don't worry if yours isn't the same as something you see in the book. As long as you get the results you want, your solution is good.

In [73]:
num=input("Give me an interger")
if ((int(num))%2==0):
    print("even")
else:
    print("odd")


Give me an interger 4


even


#### **Exercise 8**

###### sections to reference: 2.10, 3.1, 3.3, 3.4, 3.5

Another simple one! Write a program that asks the user for *two* integers. If they're equal, print something like "`5 and 5 are equal.`" If the first integer is greater than the second, print a statement with both of the integers, saying that. If the first integer is *less* than the second, print a statement saying *that*.

In [ ]:
num=input("Give me an interger")
num2=input("Give me an interger again!")
if (num==num2):
    print(num+ " is equal to" + num2)
else:
    print( num+ "=" + num2 + " IS NOT TRUE")

#### **Exercise 9**

###### section to reference: 3.7

This exercise is a little more open-ended. In a code cell below this one, rewrite one of your previous exercises that takes user input (you're probably going to want to pick from exercises 4, 6, 7, or 8). Using `try` and `except`, **make your program "idiot-proof"**. I should be able to type basically whatever I want into your input prompts without crashing your program. For example, if you ask me for a number, I should be able to type in "apples" and, instead of your program crashing and giving me a system error, I should see a message *you* typed in, like "Enter a number, you dummy." (In real life, it's probably not a good idea to insult your users, though.)

In [2]:
hours = input("How many hours do you work?")
wage = input("What is your wage?")
taxes = input("What percent of your income they pay in taxes?")
try:
    hours= (float(hours))
    wage= (float(wage))
    taxes=(float(taxes))
    gross=hours*wage
    print(gross)
    print(gross*taxes)
except ValueError:
        print("Oops!  That was no valid number.  Try again...")

    


How many hours do you work? 3
What is your wage? rc
What percent of your income they pay in taxes? cd


Oops!  That was no valid number.  Try again...


#### **Exercise 10**

###### sections to reference: 2.10, 3.1, 3.2, 3.3, 3.6, 3.7

You're going to make a program that takes scores on a scale of 0 to 100 and returns a letter grade. If the user enters a valid score, print out a letter from A to F. If the user enters invalid input, print out an error message.

Use the following table to do your conversions: 
    
Score  | Grade
-------|-------
94-100 | A
85-93  | B
74-84  | C
63-73  | D
≤62    | F
    
(Note: this *isn't* the scale we're using in this class. Just an example. :D)

Make sure you **test your program multiple times** with different inputs to make sure you handled errors and the different scores properly. You won't be graded on this, but **commenting** your code can make it easier for you to understand the many different statements in a row that you'll be writing.

In [23]:
grade=input("What did you get on the test (numerical)?")
try:
    grade=(int(grade))
    if grade>100:
        print("Lies, not possible")
    elif grade>=94:
        print("A")
    elif grade>=85:
        print("B")
    elif grade>=74:
        print("C")
    elif grade>=63:
        print("D")
    else:
        print("F")
except ValueError:
    print("Need a number!")
    

What did you get on the test (numerical)? ih


Need a number!


#### **Exercise 11**

###### sections to reference: 2.9, 2.10

Last week, you did this:

In [7]:
name = input(prompt="What is your name? ")
print("Hello", name, "!")

What is your name?  rt


Hello rt !


Run that code again, just to remind yourself. 

Your task for this week is to extend this a little bit. Copy and paste the above code into a new code cell below. You're going to make three changes to it. 

1. Make `"What is your name? "` a variable. This makes it easier to reuse and change strings that you use across larger programs. After you've done that, change the line of code you use to prompt for the person's name so that, instead of saying `prompt="What is your name? "`, it says `prompt=WHATEVER YOU NAMED YOUR PROMPT VARIABLE`. (Obviously, you'd put the name of your variable in there.)
2. Notice how, when you ran the code cell at the beginning of the exercise, it printed your name with spaces around it, like, "`Hello Sean !`"?   
Well, that doesn't look super great, so I want you to fix it. Use string *concatenation* (remember what `+` does when you use it with strings?) to insert the user's name that they enter into a string that says "Hello" to them. Then, instead of saying `print("Hello", name, "!")`, you should be able to just do `print(WHATEVER YOU NAMED YOUR HELLO VARIABLE)`. Isn't that nicer? And no ugly spaces, too!
3. One last thing: I want you to ask for the user's *age*, too. So, make another prompt where you ask "What is your age? ". Do the same thing you did in step 1, and assign that string to a variable. You want both of your prompts *before* you print anything out.  
So, now, you can change your print statement. Go ahead and concatenate the age onto the string you're printing out at the end. After you've done this, your program will go something like this:  

> `What's your name?` (you type in your name, say "Julie")  
> `How old are you?` (you type in your age, say "19")  
> `Hi, Julie! You are 19 years old.`   

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Obviously, `Julie` and `19` would be replaced with whatever you entered.

**NOTE**: This is a lot to throw at you all at once. Consider this a bit of a challenge problem, especially if this is your first time doing any serious programming. If you don't get it all at once, or something breaks, don't feel frustrated! Come see me in office hours, or send me an email. 

In [15]:
obiwan="Hello "
nameAsk="What is your name? "
ageAsk="How old are you? "
name = input(prompt=nameAsk)
obiwan="Hello " + name
age = input(prompt=ageAsk)
ageTeller= "You are "+ age+ " years old."
print(obiwan+ "!  "+ ageTeller )

What is your name?  hfg
How old are you?  34


Hello hfg!  You are 34 years old.


#### **Exercise 12**

###### sections to reference: your mind *(whoa)*

Write a little bit about this lab in a Markdown cell below. It doesn't have to be an essay, just a few sentences about your experience. Let us know if you got stuck anywhere, if you had any errors with Jupyter, if anything was boring or too easy, and if this lab was too much or not enough practice for you.

Thanks!

I enjoyed it alot! It was daunting at first and it took a bit to finish but it was a good practice tool.